Adds the `rocm-mpich-base` container.

dipietrantonio opened the pull request:
I will prepare the equivalent docker files after having done more experimenting with it.

Commit:
Adds rocm-mpich and rocm-tensorflow containers.

marcodelapierre commented on mpi/rocm-mpich-base/ubuntu22-rocm.def:
```diff
@@ -0,0 +1,66 @@
+	cd /tmp/build
+	
+	# Libfabric
+	wget https://github.com/ofiwg/libfabric/archive/refs/tags/v1.18.1.tar.gz
```
I would parametrise the Libfabric version

marcodelapierre commented on mpi/rocm-mpich-base/ubuntu22-rocm.def:
```diff
@@ -0,0 +1,66 @@
+        wget https://bootstrap.pypa.io/get-pip.py
+        python3 get-pip.py
+
+	wget https://repo.radeon.com/amdgpu-install/5.6/ubuntu/jammy/amdgpu-install_5.6.50600-1_all.deb
```
I would parametrise the version of this AMD component (driver?)

marcodelapierre commented on mpi/rocm-mpich-base/ubuntu22-rocm.def:
```diff
@@ -0,0 +1,66 @@
+
+	# AWS OFI RCCL
+	cd /tmp/build
+        git clone https://github.com/ROCmSoftwarePlatform/aws-ofi-rccl.git	
```
any way of version controlling this dependency?

marcodelapierre commented on tensorflow/rocm-tensorflow.def:
```diff
@@ -0,0 +1,25 @@
+	export HOROVOD_RCCL_HOME=/opt/rocm/rccl
+	export RCCL_INCLUDE_DIRS=/opt/rocm/rccl/include
+	export HOROVOD_RCCL_LIB=/opt/rocm/rccl/lib
+	export HCC_AMDGPU_TARGET=gfx90a
```
shall we parametrise the GPU generation?

marcodelapierre commented on tensorflow/rocm-tensorflow.def:
```diff
@@ -0,0 +1,25 @@
+	export RCCL_INCLUDE_DIRS=/opt/rocm/rccl/include
+	export HOROVOD_RCCL_LIB=/opt/rocm/rccl/lib
+	export HCC_AMDGPU_TARGET=gfx90a
+	python3 -m pip install --force-reinstall --upgrade --no-cache-dir "horovod==0.28.1"
```
I would parametrise the horovod version

marcodelapierre commented on mpi/rocm-mpich-base/ubuntu22-rocm.def:
```diff
@@ -0,0 +1,66 @@
+%post
+	ln -s /usr/share/zoneinfo/Australia/Perth /etc/localtime
+	apt -y update
+	apt -y upgrade
```
this is also consistent with other recipes in this repo - the design choice was indeed made on the rationale explained above

marcodelapierre reviewed (dismissed):
Great work!

General comment: I strongly recommend considering the migration to Dockerfiles ASAP, they are the de facto standard. We could then promptly generate SIF images after the Docker build, for HPC-ready compact images.

Both images built successfully.

`import tensorflow` in the Tensorflow image produced the following:
```
2023-08-28 06:38:58.329453: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
```
Can this be addressed in a time-effective way? Otherwise good to defer to later.

See comments across the codes.

dipietrantonio committed:
Adds PyTorch, adds dockerfiles.

dipietrantonio committed:
Updates rocm-mpich-base

AlexisEspinosaGayosso commented:
@dipietrantonio , would it be possible to separate this as three different pull requests? One for the rocm-mpich-base.dockerfile, one for the pytorch and a third for tensorflow? In that way, a fix for one will not block the others (if the others are ready).

For example, I think we are in perfect position of pulling the rocm one. Indeed, I think this is sort of "urgent". As many other images (ours, a users made) can be built FROM this one. By the way, in that one, maybe we should stick to rocm/5.6 due to driver compatibility.

In regards to the PyTorch one, I did not read carefully, but can that and also the Tensorflow ones start FROM the rocm-mpich one using dockerfiles?

Cheers.

dipietrantonio commented:
Hi @AlexisEspinosaGayosso ,
> would it be possible to separate this as three different pull requests?  
Yes I will do that, makes more sense.

> For example, I think we are in perfect position of pulling the rocm one. 

I am still working on that one, testing a few remaining things. I hope I can push it today to our Pawsey repository.

I will update this PR to be just about the `rocm-mpich-base`. I think the appropriate name should be `ubuntu:22-rocm5.6.0-mpich3.4.3` as this is mainly a base Ubuntu image, containing a few needed libraries. We should have a discussion about naming conventions.

dipietrantonio committed:
Removes pytorch and tensorflow - will create separate PRs for them.

dipietrantonio committed:
Reverts to ROCm 5.6

The reason is the driver on Setonix is too old for ROCm 5.7.

Also set a few important environment variables.

dipietrantonio committed:
Whitespaces.

dipietrantonio commented:
@pelahi @AlexisEspinosaGayosso this PR is ready for review.

dipietrantonio committed:
Adds OSU benchmarks and sets ENVs.

pelahi reviewed (commented):
Is there a reason why this container doesn't build from the mpich lustre container?

dipietrantonio commented:
I install libfabric as a dependency of https://github.com/ROCmSoftwarePlatform/aws-ofi-rccl , which is needed for optimal performance of the ROCm Communication Collective Library (RCCL). So this is a AMD specific build. I might get away by installing lib fabric and aws-ofi-rccl on top of the mpich-lustre container. I will try.

Commit:
Update to mpich rocm recipe

- added a readme to describe how to build the container
- update the recipe to add luster to ensure lustre aware mpi
- updated recipe to generalise it a bit, added comments to

Commit:
updated defaults, renamed readme

Commit:
fixed additional command necessary prior to rocm PR
